### PR TITLE
`ultralytics 8.3.203` Restore `GradScaler` on resuming training

### LINKED ISSUE #17282: Losses are nan upon resuming training with resume=True [Yolov8]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

I am facing an issue where upon resuming an interrupted training (manual interruption after 51 completed epochs) the losses are all nan from the very beginning. 

I had created epochs on every 5 checkpoints, so i'm loading last.pt (also tried epoch50.pt) but to no avail. The weird thing is that If I begin a new training from the last checkpoint without the "resume" parameter, it works just fine. 

However, setting the device="cpu" resumes the training successfully with resume=True. This obviously is not a solution, as my dataset is too large to even consider using CPU.

It is worth noting i have many existing solutions, reducing batch size down to 2, changing amp to false, directly modifying the args.yml file, etc. Tried setting lr0 to 0.001. Also worth mentioning I have been checking many solutions online, but most people seem to have an issue with regular training, but my training only has this issue upon using "resume". 

Please suggest some solutions. 

### Environment

Ultralytics 8.3.19 🚀 Python-3.10.15 torch-2.5.0+cu121 CUDA:0 (NVIDIA GeForce RTX 4090, 24564MiB)

Using Yolov8x.

All AMP checks are passed, and everything else is normal. I can run other new trainings without any issues. 

### Minimal Reproducible Example

Below is a simple example of what my code looks like. It is worth noting i have tried a dozen variations of this, reducing batch size down to 2, changing amp to false, directly modifying the args.yml file, etc. Also tried different checkpoints. 

=========================

```
def train_yolov8(model_path, **kwargs):
    model = YOLO(model_path)
    model.train(**kwargs)

if __name__ == '__main__':
    epochs = 300
    train_yolov8(
        model_path=r"my/model/full/path",
        resume=True
    )
```

### Additional

I am using the LVIS dataset. 

Furthermore, I have faced this issue once a week or two ago as well. So it is not a one-time issue.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Fixes https://github.com/ultralytics/ultralytics/issues/21913
Fixes https://github.com/ultralytics/ultralytics/issues/17282

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds AMP scaler state saving/loading to improve reliability when resuming mixed-precision training, plus a version bump to 8.3.203. ⚙️🚀

### 📊 Key Changes
- Bumped version: `8.3.202` → `8.3.203`.
- Trainer now saves the AMP GradScaler state in checkpoints: `"scaler": self.scaler.state_dict()`.
- Trainer restores the scaler state on resume if present: `self.scaler.load_state_dict(ckpt["scaler"])`.
- Minor cleanup in checkpoint loading checks.

### 🎯 Purpose & Impact
- More reliable resume for mixed-precision (AMP) training ✅
  - Preserves loss scaling state to avoid instability or slow warm-up after resuming.
- Backward compatible with older checkpoints 🙌
  - Scaler is loaded only if available; older checkpoints continue to work.
- Smoother training continuity and potentially fewer NaNs/overflows on resume 💪

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response — an Ultralytics engineer will review and assist soon. In the meantime, please review the checklist below to help us streamline merging:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional suggestions for this PR:
- Please add or confirm tests that:
  - Save and resume with AMP enabled to verify GradScaler state is restored.
  - Resume from older checkpoints that lack a scaler entry to ensure backward compatibility.
  - Handle non-AMP/CPU-only runs without errors when no scaler is present.
- Confirm that the new checkpoint keys are documented where appropriate in [documentation](https://docs.ultralytics.com/) and that any user-facing behavior on resume is clear.
- Double-check that all paths touching resume logic pass CI, including mixed-precision configurations ⚡

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🎉

**Comment by glenn-jocher** [comment]
@Y-T-G what does this do to the checkpoint sizes?

**Comment by Y-T-G** [comment]
@glenn-jocher It doesn't affect much:

**main**
```
ls -lh runs/detect/train/weights/
total 32M
-rw-r--r-- 1 root root 16M Sep 22 13:15 best.pt
-rw-r--r-- 1 root root 16M Sep 22 13:15 last.pt
```

**PR**
```
ls -lh runs/detect/train/weights/
total 32M
-rw-r--r-- 1 root root 16M Sep 22 13:11 best.pt
-rw-r--r-- 1 root root 16M Sep 22 13:11 last.pt
```

It only stores a few values:
```python
In [4]: model["scaler"]
Out[4]: 
{'scale': 2048.0,
 'growth_factor': 2.0,
 'backoff_factor': 0.5,
 'growth_interval': 2000,
 '_growth_tracker': 8}
```

**Comment by glenn-jocher** [comment]
@Y-T-G oh that's good!

**Comment by UltralyticsAssistant** [comment]
🎉 Big cheers on the merge of [PR #22189](https://github.com/ultralytics/ultralytics/pull/22189)! 

“Great things are done by a series of small things brought together.” — Vincent van Gogh. This thoughtful addition to save/load the AMP GradScaler state is one of those precise improvements that elevates reliability across the board—smoother resumes, fewer NaNs, and better continuity for mixed-precision training. 

Huge thanks to @Y-T-G for the implementation and to @glenn-jocher for the collaboration. Also appreciate the clean version bump to 8.3.203 and the backward-compatible design—exactly the kind of polish that keeps our users confident and productive.

Quick tip: resume seamlessly with `yolo train model=yolo11n.pt data=coco8.yaml resume=True`.

Appreciate you both for pushing the Ultralytics ecosystem forward! 🚀
